### PR TITLE
assumptions: Fix handlers for Mul of 0 with irrational/imaginary numbers

### PR DESCRIPTION
**Pull request by SujalKumar06**
#### References to other Issues or PRs

Fixes: #29480

#### Brief description of what is fixed or changed
`ask` was not properly handling multiplication with 0 for irrational/imaginary numbers.

The issue was with handlers of `Integer`, `Real` and `Imaginary` predicates, completely skipping over the edge case being the `expr` could be 0.

New tests have been added in `test_query.py` while some tests are updated following the new correct behaviour.

#### Release Notes

* assumptions
  * Fixed a bug where `ask` used to give wrong results when 0 was multiplied by irrational/imaginary numbers. For example - `ask(Q.real(n * I), Q.zero(n))` returns `True` now instead of wrongly returning `False`.

### CODE REVIEW

File: sympy/assumptions/tests/test_query.py [hunk 1]
```diff
@@ -1652,6 +1652,8 @@ def test_imaginary():
     # issue 7886
     assert ask(Q.imaginary(Pow(x, Rational(1, 4))), Q.real(x) & Q.negative(x)) is False
 
+    # https://github.com/sympy/sympy/issues/29480
+    assert ask(Q.imaginary(n * I), Q.zero(n)) is False
```
Comment (TiloRC) [inline sympy/assumptions/tests/test_query.py]: Zero is considered imaginary so this should give back True, not False.

File: sympy/assumptions/handlers/sets.py [hunk 2]
```diff
@@ -101,6 +104,8 @@ def _(expr, assumptions):
             else:
                 return
 
+    if _output is False:
+        return False if is_zero is False else None
```
Comment (TiloRC) [inline sympy/assumptions/handlers/sets.py:108]: In the  `Odd/Even -> !Integer` case, we would give back None here when it would be fine to give False.

Comment (SujalKumar06) [reply sympy/assumptions/handlers/sets.py:108]: I am not sure if this is causing an issue here, the place where `odd/even` is handled is 
```python
if arg.is_Rational:
    if arg.q == 2:
        return ask(Q.even(2*expr), assumptions)
    if ~(arg.q & 1):
        return None
```
`print(ask(Q.integer(n/x), Q.even(x) & Q.odd(n)))` This still returns None even after commenting the logic.

File: sympy/assumptions/tests/test_query.py (hunk 1 above)
Comment (SujalKumar06) [reply sympy/assumptions/tests/test_query.py]: Yeah I rectified in the issue but forgot to update it here. I will do it

**Comment by SujalKumar06** [comment]
Hey @TiloRC ,
I have pushed the changes according to reviews.  You can have a look over them.

**Comment by SujalKumar06** [comment]
Hey @TiloRC ,
Can you have a look over this?

**Review by JosephMehdiyev — COMMENTED** [review]
If I am understanding correctly, this PR should use the future recursive ask handler from this PR https://github.com/sympy/sympy/pull/29921 if it is merged, otherwise these ask calls should be computationally expensive
Other than that changes seem OK to me

**Comment by TiloRC** [comment]
If you want, you can take over this PR and finish it assuming Sujal is ok with that (and I would expect he would be). I would recommend you wait until https://github.com/sympy/sympy/pull/29921 gets merged to work on this though so it's easier.

**Comment by JosephMehdiyev** [comment]
> If you want, you can take over this PR and finish it assuming Sujal is ok with that (and I would expect he would be).

Sure, I am open to that

**Comment by SujalKumar06** [comment]
> If you want, you can take over this PR and finish it assuming Sujal is ok with that 

Sorry I was busy handling some other PRs, sure @JosephMehdiyev continue working on the PR, I don't have any issues with it!!

I will close this one for now.
